src/authentication.cpp

```diff
@@ -0,0 +1,239 @@
+                // except for this.
+                s << '%' << std::uppercase << std::hex << std::setw(2) << std::setfill('0') << (int) c;
+                s.copyfmt(state);
+            }
+            else {
+                s << c;
+            }
+        }
+        return s.str();
+    }
+
+    std::string get_user_signing_key(
+        const std::string_view& secret_key,
+        const std::string_view& date,
+        const std::string_view& region)
+    {
+        // Hate it when amazon gives me homework
+        // during implementation of their protocol
+        std::cout << "date time component is " << date << std::endl;
+        auto date_key = irods::s3::authentication::hmac_sha_256(std::string("AWS4").append(secret_key), date);
+        auto date_region_key = irods::s3::authentication::hmac_sha_256(date_key, region);
+        // 'date region service key'
+        // :eyeroll:
+        auto date_region_service_key = irods::s3::authentication::hmac_sha_256(date_region_key, "s3");
+        return irods::s3::authentication::hmac_sha_256(date_region_service_key, "aws4_request");
+    }
+
+    // TODO when this is working, we can improve performance by reusing the same stringstream where possible.
+
+    // Turn the url into the 'canon form'
+    std::string canonicalize_url(const static_buffer_request_parser& request, const boost::urls::url_view& url)
+    {
+        std::stringstream result;
+        // result << request.get().at("Host");
+        for (auto i : url.segments()) {
+            result << '/' << uri_encode(i); // :shrug:
+        }
+        return result.str();
+    }
+
+    bool should_include_header(const std::string_view& sv)
+    {
+        return sv.starts_with("X-Amz") || sv.starts_with("x-amz") || sv == "Host" || sv == "Content-Type" ||
+               sv == "Content-MD5";
+    }
+
+    std::string canonicalize_request(
+        const static_buffer_request_parser& request,
+        const boost::urls::url_view& url,
+        const std::vector<std::string>& signed_headers)
+    {
+        // At various points it wants various fields to be sorted.
+        // so reusing this can at least avoid some of the duplicate allocations and such
+        std::vector<std::string_view> sorted_fields;
+
+        std::stringstream result;
+
+        std::ios state(nullptr);
+        state.copyfmt(result);
+
+        result << request.get().method_string() << '\n';
+        result.copyfmt(result); // re store former formatting
+        result << canonicalize_url(request, url) << '\n';
+        // Canonicalize query string
+        {
+            bool first = true;
+            // Changing it to a pair enables us to sort easily.
+            std::vector<std::pair<std::string, std::string>> params;
+            std::transform(
+                url.encoded_params().begin(), url.encoded_params().end(), std::back_inserter(params), [](auto a) {
```

Comment:
Should `a` be taken as a copy or by a reference?

Reply:
Good catch